R/common-impact-coding.R

```diff
@@ -0,0 +1,179 @@
+#'                                        replace = TRUE))
+#'                
+#' # Continuous target and one high_cardinality variable
+#' pphcv(df = dat, target = "Ycontinuous", high_card = "hcX", 
+#'       type = "continuous")
+#'           
+#' # Binary target and one high_cardinality variable
+#' pphcv(df = dat, target = "Ycategorical", high_card = "hcX", pos_class = "Y", 
+#'       type = "binary")
+#'           
+#' # Binary target coded as numeric(0, 1) and one high-cardinality variable 
+#' # and specifying the pos_class.
+#' pphcv(df = dat, target = "YCat01", high_card = "hcX", pos_class = 1, 
+#'       type = "binary")
+#'           
+#' # Binary target coded as numeric(0, 1) and one high-cardinality variable and 
+#' # not specifying the pos_class. Use type = "continuous". Should give same
+#' # results as if type is binary and positive class is specified.
+#' pphcv(df = dat, target = "YCat01",  high_card = "hcX", type = "continuous", 
+#'       m = 1)
+#' 
+#' # Multiclass target and one high-cardinality variable
+#' pphcv(df = dat, target = "Ymulticlass", high_card = "hcX", 
+#'       type = "multiclass")
+#'           
+#' # Binary target and a vector of high-cardinality variables.
+#' vec <- c("hcX", "hcX2", "hcX3")
+#' pphcv(df = dat, target = "Ycategorical", high_card = vec, 
+#'           type = "binary", pos_class = "Y")
+#' 
+#' # Multiclass target and a vector of high-cardinality variables and specify
+#' # remove_high_card = TRUE
+#' vec <- c("hcX", "hcX2", "hcX3")
+#' pphcv(df = dat, target = "Ymulticlass", high_card = vec, type = "multiclass", 
+#'       remove_high_card = TRUE)
+pphcv <- function(df, target, high_card, type, pos_class = NULL, m = NULL,
+                  multi_base = NULL, remove_high_card = FALSE) {
+  
+  for (i in high_card) {
+    # For binary, if target is numeric, pos_class must also be numeric.
+    if (type == "binary" && is.numeric(df[[target]]) && 
+        !is.numeric(pos_class)) {
+      stop("If specified binary target is numeric, pos_Class must also be 
+           numeric. You may also use type = 'continuous' with no specified 
+           pos_class")
+    }
+    # For binary, must have a pos_class specified
+    if (type == "binary" && is.null(pos_class)) {
+      stop("You must specify the positive class of your binary target")
+    }
+    # If any NAs are present, throw error that user must deal with them 
+    # beforehand.
+    if (anyNA(df[[target]]) || anyNA(df[[i]])) {
+      stop("You must deal with NAs before reducing cardinality. Consider 
+           re-coding all NAs in high_card into a new category 'missing'. 
+           Also consider removingrows where target is missing.")
+    }
+    # If type is multiclass and the target is numeric, give a warning about
+    # coercion to factor.
+    if (type == "multiclass" && is.numeric(df[[target]])) {
+      warning("Your specified target is multiclass and numeric. It was coerced
+              into a factor for use.")
+    }
```

Comment:
Would you put code somewhere in here to check user input for `target`, `pos_class`, `high_card`?

1. Check that the column in the function argument is actually in the df? For `pos_class`, check that value is in the column values. Could use `unique` function. If it's not found, throw error with helpful message...

2. Write an associated unit tests?

I ran these scenarios and there was either no error or no helpful error. Thanks!